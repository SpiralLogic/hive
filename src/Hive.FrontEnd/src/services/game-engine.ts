import { GameId, GameState, Move, PlayerId } from '../domain';
import { AiMode, HexEngine } from '../domain/engine';
import { getAllPlayerTiles } from '../utilities/hextille';

export default class GameEngine implements HexEngine {
  public currentPlayer: PlayerId;
  public initialGame: Promise<GameState>;
  public gameId: GameId;
  private aiMode: AiMode;
  private currentRequest: Promise<GameState>;
  private requestHeaders = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
  };

  constructor(existingGame?: { gameId: string; currentPlayer: string }) {
    this.currentPlayer = Number(existingGame?.currentPlayer ?? 0);
    this.gameId = existingGame?.gameId ?? '';
    this.aiMode = this.currentPlayer === 0 ? 'on' : 'off';

    this.currentRequest = existingGame?.gameId
      ? this.getExistingGame(existingGame.gameId)
      : this.getNewGame();

    this.initialGame = this.completeRequest();
  }

  setAiMode = async (mode: AiMode) => {
    this.aiMode = mode;
    if (mode === 'on') {
      const { cells, players } = await this.completeRequest();
      if (
        getAllPlayerTiles(this.currentPlayer, players, cells).reduce(
          (count, tile) => count + tile.moves.length,
          0
        ) === 0
      ) {
        await this.aiMove();
      }
    }
    await this.completeRequest();
  };

  getAiMode = () => this.aiMode;

  move = async (move: Move): Promise<GameState> => {
    await this.postRequest(`/api/move/${this.gameId}`, move);
    if (this.aiMode === 'on') {
      await this.completeRequest();
      return this.aiMove();
    }
    return this.completeRequest();
  };

  private aiMove = async (): Promise<GameState> => {
    await this.postRequest(`/api/ai-move/${this.gameId}/${this.currentPlayer == 0 ? 1 : 0}`);
    return this.completeRequest();
  };

  private completeRequest = async () => {
    const response = await this.currentRequest;
    if (!this.gameId) this.gameId = response.gameId;
    return response;
  };

  private getExistingGame = async (gameId: GameId) => {
    return this.getRequest(`/api/game/${gameId}`);
  };

  private getNewGame = async () => {
    return this.postRequest(`/api/new`);
  };

  private postRequest = async (url: string, body?: Move) => {
    const response = await fetch(url, {
      method: 'POST',
      headers: this.requestHeaders,
      body: body ? JSON.stringify(body) : undefined,
    });
    this.currentRequest = response.json() as Promise<GameState>;
    return this.currentRequest;
  };

  private getRequest = async (url: string) => {
    const response = await fetch(url, {
      method: `GET`,
      headers: this.requestHeaders,
    });

    this.currentRequest = response.json() as Promise<GameState>;
    return this.currentRequest;
  };
}
