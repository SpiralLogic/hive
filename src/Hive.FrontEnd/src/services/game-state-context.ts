import {Cells, GameState, GameStatus, HexCoordinates, Player, Players, TileMapKey} from '../domain';
import {batch, signal} from '@preact/signals';
import {HistoricalMoves} from '../domain/historical-move';
import {useContext} from 'preact/hooks';
import {createContext} from 'preact';

type NonEmptyArray<T> = [T, ...T[]];

function isNonEmpty<A>(array: Array<A>): array is NonEmptyArray<A> {
    return array.length > 0;
}

export const moveMap = signal(new Map<TileMapKey, HexCoordinates>());

const updateMoveMap = (players: Player[], cells: Cells) => {
    const nextMoveMap = new Map<TileMapKey, HexCoordinates>();
    for (const t of players.flatMap((p) => p.tiles)) {
        nextMoveMap.set(`${t.playerId}-${t.id}`, t.moves);
    }

    for (const t of cells.flatMap((c) => c.tiles).filter((t) => isNonEmpty(t.moves))) nextMoveMap.set(`${t.playerId}-${t.id}`, t.moves);
    moveMap.value = nextMoveMap;
};

const createGameState = () => {
    const cells = signal<Cells>([]);
    const players = signal<Players>([]);
    const history = signal<HistoricalMoves>([]);
    const gameStatus = signal<GameStatus>('NewGame');
    const gameId = signal<string>('');
    const setGameState = (gameState: GameState) => {
        if (gameState.history.length === 0 || gameState.history.length > history.peek().length) {
            updateMoveMap(gameState.players, gameState.cells);
            batch(() => {
                cells.value = gameState.cells;
                if (
                    gameState.players.some(
                        (p) => p.tiles.length !== players.peek().find((p2) => p.id === p2.id)?.tiles.length
                    )
                )
                    players.value = gameState.players;
                history.value = gameState.history;
                gameStatus.value = gameState.gameStatus;
                gameId.value = gameState.gameId;
            });
        }
    };

    return {cells, players, history, gameStatus, gameId, setGameState};
};
export const GameStateContext = createContext(createGameState());
export const useGameState = () => useContext(GameStateContext);
