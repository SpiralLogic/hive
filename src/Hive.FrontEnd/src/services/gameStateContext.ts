import { Cells,  GameState, GameStatus, HexCoordinates, Player, Players, TileMapKey } from '../domain';
import { batch, signal } from '@preact/signals';
import { HistoricalMoves } from '../domain/historical-move';
import { useContext } from 'preact/hooks';
import { createContext } from 'preact';

type NonEmptyArray<T> = [T, ...T[]];

function isNonEmpty<A>(arr: Array<A>): arr is NonEmptyArray<A> {
  return arr.length > 0;
}
export const moveMap = signal(new Map<TileMapKey, HexCoordinates>());

const updateMoveMap = (players: Player[], cells: Cells) => {
  const newMoveMap = new Map<TileMapKey, HexCoordinates>();
  players
    .flatMap((p) => p.tiles)
    .forEach((t) => {
      newMoveMap.set(`${t.playerId}-${t.id}`, t.moves);
    });

  cells
    .flatMap((c) => c.tiles)
    .filter((t) => isNonEmpty(t.moves))
    .forEach((t) => newMoveMap.set(`${t.playerId}-${t.id}`, t.moves));
  moveMap.value = newMoveMap;
};

const createGameState = () => {
  const cells = signal<Cells>([]);
  const players = signal<Players>([]);
  const history = signal<HistoricalMoves>([]);
  const gameStatus = signal<GameStatus>('NewGame');
  const gameId = signal<string>('');
  const setGameState = (gameState: GameState) => {
    if (gameState.history.length === 0 || gameState.history.length > history.peek().length) {
      updateMoveMap(gameState.players, gameState.cells);
      batch(() => {
        cells.value = gameState.cells;
        if (
          gameState.players.some(
            (p) => p.tiles.length !== players.peek().find((p2) => p.id === p2.id)?.tiles.length
          )
        )
          players.value = gameState.players;
        history.value = gameState.history;
        gameStatus.value = gameState.gameStatus;
        gameId.value = gameState.gameId;
      });
    }
  };

  return { cells, players, history, gameStatus, gameId, setGameState };
};
export const GameStateContext = createContext(createGameState());
export const useGameState = () => useContext(GameStateContext);
