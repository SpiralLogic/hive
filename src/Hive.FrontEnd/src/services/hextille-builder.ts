import { Cell, Row } from '../domain';

export class HextilleBuilder {
  private readonly hexagons: Array<Cell>;
  private readonly firstRow: number;
  private readonly height: number;
  private readonly firstColumn: number;
  private readonly width: number;

  constructor(hexagons: Array<Cell>) {
    hexagons.sort((c1, c2) => c1.coords.r - c2.coords.r || c1.coords.q - c2.coords.q);
    [this.firstRow, this.height] = this.determineHeight(hexagons);
    [this.firstColumn, this.width] = this.determineWidth(hexagons);
    this.hexagons = hexagons;
  }

  createRows = (): Array<Row> => {
    const rows = this.createEmptyRows();
    for (const cell of this.hexagons) {
      rows[cell.coords.r - this.firstRow].cells[cell.coords.q - this.firstColumn] = cell;
    }
    return rows;
  };
  private determineWidth = (cells: Array<Cell>): [number, number] => {
    const min = Math.min(...cells.map((c) => c.coords.q));
    const max = Math.max(...cells.map((c) => c.coords.q));
    return [min, max - min + 1];
  };

  private determineHeight = (cells: Array<Cell>): [number, number] => {
    const firstCell = cells[0];
    const lastCell = cells[cells.length - 1];
    const height = lastCell.coords.r - firstCell.coords.r + 1;
    return [firstCell.coords.r - 1, height + 2];
  };

  private createEmptyRow = (index: number): Row => {
    return {
      id: this.firstRow + index,
      hidden: index === 0 || index === this.height - 1,
      cells: [...Array.from({ length: this.width }).keys()].map((index_: number) => ({
        coords: { q: this.firstColumn + index_, r: this.firstRow + index },
        tiles: [],
        hidden: true,
      })),
    };
  };

  private createEmptyRows = () =>
    [...Array.from({ length: this.height }).keys()].map((element) => this.createEmptyRow(element));
}
