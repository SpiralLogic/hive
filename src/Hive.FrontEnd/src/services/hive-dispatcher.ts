import { Move, PlayerId, Tile } from '../domain';

export type HiveEventListener<T extends HiveIntent> = <U extends T>(intent: U) => void;
export type HiveIntent = HiveEvent | HiveAction;
export type HiveEvent = MoveEvent | TileEvent | ConnectEvent;
export type HiveAction = Action | TileAction;
export type MoveEvent = { type: 'move'; move: Move };
export type Action = {
  type: 'tileClear';
};
export type ConnectEvent = { type: 'opponentConnected' | 'opponentDisconnected'; playerId: PlayerId };
export type TileEvent = {
  type: 'tileDropped' | 'click' | 'tileSelected' | 'tileDeselected';
  tile: Omit<Tile, 'moves'>;
  fromEvent?: boolean;
};
export type TileAction = {
  type: 'tileSelect' | 'tileDeselect';
  tile: Omit<Tile, 'moves'>;
};

export class HiveDispatcher {
  private listeners = new Map<string, Set<HiveEventListener<HiveIntent>>>();

  add = <T extends HiveIntent>(type: T['type'], listener: HiveEventListener<T>): (() => void) => {
    if (!this.listeners.get(type)?.add(listener)) this.listeners.set(type, new Set([listener]));
    return () => this.remove<T>(type, listener);
  };

  dispatch = <T extends HiveIntent>(intent: T): void => {
    const listeners = this.listeners.get(intent.type) ?? [];
    for (const l of listeners) {
      l(intent);
    }
  };

  remove = <T extends HiveIntent>(type: T['type'], listener: HiveEventListener<T>): void => {
    this.listeners.get(type)?.delete(listener);
  };
}
