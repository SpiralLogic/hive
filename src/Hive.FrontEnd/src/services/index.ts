export * from './hextille-builder';
export * from './hive-dispatcher';
export * from './server-connection';
