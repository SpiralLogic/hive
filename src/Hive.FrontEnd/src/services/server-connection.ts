import {
  HttpTransportType,
  HubConnection,
  HubConnectionBuilder,
  HubConnectionState,
} from '@microsoft/signalr';

import { GameId, GameState, PlayerId, Tile } from '../domain';

export type PlayerSelectionEvent = 'select' | 'deselect';
export type PlayerConnectionEvent = 'connect' | 'disconnect';
export type GameStateUpdateHandler = (gameState: GameState) => void;
export type OpponentSelectionHandler = (type: PlayerSelectionEvent, tile: Tile) => void;
export type OpponentConnectedHandler = (type: PlayerConnectionEvent) => void;
export type ServerConnectionConfig = {
  currentPlayer: PlayerId;
  gameId: GameId;
  updateHandler: GameStateUpdateHandler;
  opponentSelectionHandler: OpponentSelectionHandler;
  opponentConnectedHandler: OpponentConnectedHandler;
};
export interface ServerConnection {
  connectGame: () => Promise<void>;
  getConnectionState: () => HubConnectionState;
  sendSelection: OpponentSelectionHandler;
  closeConnection: () => Promise<void>;
}
class ServerConnectionImpl implements ServerConnection {
  private readonly updateHandler;

  private readonly gameId: GameId;

  private readonly connection: HubConnection;

  private readonly opponentSelectionHandler: OpponentSelectionHandler;

  private readonly opponentConnectedHandler: OpponentConnectedHandler;

  private readonly currentPlayer: PlayerId;

  constructor(config: ServerConnectionConfig) {
    const { currentPlayer, gameId, updateHandler, opponentSelectionHandler, opponentConnectedHandler } =
      config;
    this.currentPlayer = currentPlayer;
    this.gameId = gameId;
    this.opponentSelectionHandler = opponentSelectionHandler;
    this.opponentConnectedHandler = opponentConnectedHandler;
    this.updateHandler = updateHandler;
    this.connection = this.createConnection();
  }

  connectGame = (): Promise<void> => {
    if (!import.meta.env.PROD) {
      this.connection.onreconnecting((error) =>
        console.warn(`reconnecting to game ${this.gameId} .. ${error}`)
      );
      this.connection.onclose((error) =>
        console.info(`connection closed to game ${this.gameId} .. ${error}`)
      );
      this.connection.onreconnected((error) => {
        window.location.reload();
        console.info(`reconnected to game ${this.gameId} .. ${error}`);
      });
    }
    return this.connection.start();
  };

  getConnectionState = (): HubConnectionState => this.connection.state;

  sendSelection: OpponentSelectionHandler = (type: 'select' | 'deselect', tile: Tile) => {
    if (this.connection.state === HubConnectionState.Connected)
      this.connection
        .send('SendSelection', type, tile)
        .catch((error: Error) => console.error(error.message.toString()));
  };

  closeConnection = (): Promise<void> => {
    this.connection.off('ReceiveGameState', this.updateHandler);
    return this.connection.stop();
  };

  private createConnection = () => {
    const hubUrl = `${window.location.protocol}//${window.location.host}/gamehub/${this.gameId}/${this.currentPlayer}`;
    const connection = new HubConnectionBuilder()
      .withUrl(hubUrl, {
        skipNegotiation: true,
        transport: HttpTransportType.WebSockets,
      })
      .withAutomaticReconnect()
      .build();
    connection.on('ReceiveGameState', this.updateHandler);
    connection.on('OpponentSelection', this.opponentSelectionHandler);
    connection.on('PlayerConnection', this.opponentConnectedHandler);
    return connection;
  };
}
export type ServerConnectionFactory = (config: ServerConnectionConfig) => ServerConnection;
export const serverConnectionFactory = (config: ServerConnectionConfig) => new ServerConnectionImpl(config);
