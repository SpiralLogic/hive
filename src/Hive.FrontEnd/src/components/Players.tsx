import '../css/players.css';

import { FunctionComponent } from 'preact';

import { PlayerId } from '../domain';
import GameTile from './GameTile';
import Player from './Player';
import { useGameState } from '../services/signals';

const Players: FunctionComponent<{ currentPlayer: PlayerId }> = (properties) => {
  const { currentPlayer } = properties;
  const { players } = useGameState();
  return (
    <aside className="players" aria-label={'PLayer pieces not yet on the board'}>
      {players.value.map((player) => (
        <Player key={`${player.id}`} name={player.name} id={player.id}>
          {player.tiles
            .map((tile) => ({
              ...tile,
              moves: undefined,
            }))
            .map((tile) => (
              <GameTile currentPlayer={currentPlayer} key={`${tile.id}`} {...tile} />
            ))}
        </Player>
      ))}
    </aside>
  );
};

Players.displayName = 'Players';
export default Players;
