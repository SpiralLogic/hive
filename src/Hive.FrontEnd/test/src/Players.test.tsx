import { render, screen } from '@testing-library/preact';
import Players from '../../src/components/Players';
import { GameStateContext } from '../../src/services/signals';
import { signal } from '@preact/signals';

const ant = { id: 1, playerId: 1, creature: 'ant', moves: [{ q: 1, r: 1 }] };
const fly = { id: 2, playerId: 0, creature: 'fly', moves: [] };

const defaultProps = {
  currentPlayer: 0,
};
const setup = () =>
  render(
    <GameStateContext.Provider
      value={{
        history: signal([]),
        gameId: signal('33'),
        cells: signal([]),
        gameStatus: signal('NewGame'),
        setGameState: vi.fn(),
        players: signal([
          { id: 1, name: 'Player 1', tiles: [ant, fly, fly] },
          { id: 2, name: 'Player 2', tiles: [ant] },
        ]),
      }}>
      <Players {...defaultProps} />
    </GameStateContext.Provider>
  );
describe('<Players>', () => {
  it('players are rendered', () => {
    setup();
    expect(screen.getAllByRole('heading', { name: /Player \d+/ })).toHaveLength(2);
  });

  it('renders', () => {
    global.window.history.replaceState({}, global.document.title, `/game/33/1`);
    const view = setup();
    expect(view.baseElement).toMatchSnapshot();
  });
});
