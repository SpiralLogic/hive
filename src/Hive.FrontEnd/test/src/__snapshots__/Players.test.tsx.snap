// Vitest Snapshot v1

exports[`<Players> > renders 1`] = `
<body>
  <div>
    <aside
      aria-label="PLayer pieces  not yet on the board"
      class="players"
    >
      <section
        aria-label="Player 1's unplaced pieces"
        class="player player1"
      >
        <h2>
          Player 1
        </h2>
        <div
          class="tiles"
        >
          <div
            class="player1 hex tile ant"
            draggable="false"
            title="Player-1 ant"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#ant"
              />
            </svg>
          </div>
          <div
            class="player0 hex tile fly"
            draggable="false"
            title="Player-0 fly"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#fly"
              />
            </svg>
          </div>
          <div
            class="player0 hex tile fly"
            draggable="false"
            title="Player-0 fly"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#fly"
              />
            </svg>
          </div>
        </div>
      </section>
      <section
        aria-label="Player 2's unplaced pieces"
        class="player player2"
      >
        <h2>
          Player 2
        </h2>
        <div
          class="tiles"
        >
          <div
            class="player1 hex tile ant"
            draggable="false"
            title="Player-1 ant"
          >
            <svg
              viewBox="0 0 100 100"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                href="#hex"
              />
              <use
                class="creature"
                href="#ant"
              />
            </svg>
          </div>
        </div>
      </section>
    </aside>
  </div>
</body>
`;
